### PR TITLE
THRIFT-5509: Close connection in IsOpen

### PR DESCRIPTION
**Pull request by fishy**
Client: go

When the connectivity check failed in IsOpen, close the connection
explicitly to avoid connection leaks.

This is Path 2 of THRIFT-5509.

### CODE REVIEW

**Comment by fishy** [comment]
NOTE: Only one of https://github.com/apache/thrift/pull/2513 and https://github.com/apache/thrift/pull/2514 should be merged, not both.

File: lib/go/thrift/socket_conn.go [hunk 1]
```diff
@@ -83,7 +84,17 @@ func (sc *socketConn) IsOpen() bool {
 	if !sc.isValid() {
 		return false
 	}
-	return sc.checkConn() == nil
+	if err := sc.checkConn(); err != nil {
+		if !errors.Is(err, net.ErrClosed) {
+			// The connectivity check failed and the error is not
+			// that the conncetion is already closed, we need to
```
Comment (dcelasun) [inline lib/go/thrift/socket_conn.go:90]:
```suggestion
			// that the connection is already closed, we need to
```

**Review by dcelasun — APPROVED** [review]
LGTM

**Comment by fishy** [comment]
We are testing an [equivalent fix](https://github.com/reddit/baseplate.go/pull/477) internally (on our client pool management logic). So far the test shows that it indeed fixes the leak, but we'll keep it running for the day and merge both fixes tomorrow morning.

**Comment by fishy** [comment]
Merging this in now. Will also cherry-pick the fix into 0.16.0 branch.

**Pull request merged**